File: go-helper/internal/taskagent/workspace_config.go
package taskagent

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

const workspaceConfigFileName = "clock-p.workspace.json"

type WorkspaceConfig struct {
	PlaceID        string `json:"place_id"`
	CodeSyncConfig string `json:"code_sync_config,omitempty"`
}

func LoadWorkspaceConfig(workspace string) (WorkspaceConfig, error) {
	workspaceAbs, err := filepath.Abs(workspace)
	if err != nil {
		return WorkspaceConfig{}, err
	}
	path := filepath.Join(workspaceAbs, workspaceConfigFileName)
	body, err := os.ReadFile(path)
	if err != nil {
		return WorkspaceConfig{}, fmt.Errorf("read %s: %w", path, err)
	}
	var config WorkspaceConfig
	if err := json.Unmarshal(body, &config); err != nil {
		return WorkspaceConfig{}, fmt.Errorf("parse %s: %w", path, err)
	}
	config.PlaceID = strings.TrimSpace(config.PlaceID)
	if config.PlaceID == "" {
		return WorkspaceConfig{}, fmt.Errorf("%s place_id is required", path)
	}
	if !regexp.MustCompile(`^[0-9]+$`).MatchString(config.PlaceID) {
		return WorkspaceConfig{}, fmt.Errorf("%s place_id must contain digits only", path)
	}
	config.CodeSyncConfig = normalizeWorkspaceConfigPath(config.CodeSyncConfig, "code-sync.tree.json")
	return config, nil
}

func ValidateWorkspaceBindingFiles(workspace string, config WorkspaceConfig) error {
	workspaceAbs, err := filepath.Abs(workspace)
	if err != nil {
		return err
	}
	configPath := filepath.Join(workspaceAbs, filepath.FromSlash(config.CodeSyncConfig))
	if _, err := os.Stat(configPath); err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return fmt.Errorf(
				"missing %s; clock-p.workspace.json code_sync_config points here. This file declares the Studio node tree managed by code-sync",
				configPath,
			)
		}
		return fmt.Errorf("stat %s: %w", configPath, err)
	}
	return nil
}

func normalizeWorkspaceConfigPath(value string, defaultValue string) string {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return defaultValue
	}
	return filepath.ToSlash(filepath.Clean(trimmed))
}

File: tools/bridge2/clockp_bridge2/workspace_config.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

from .errors import BridgeError


@dataclass(frozen=True)
class WorkspaceConfig:
    place_id: str
    code_sync_config: str


def load_workspace_config(workspace: Path) -> WorkspaceConfig:
    path = workspace / "clock-p.workspace.json"
    try:
        body = path.read_text(encoding="utf-8")
    except FileNotFoundError as exc:
        raise BridgeError("workspace_config_missing", "missing clock-p.workspace.json", {"path": str(path)}) from exc
    except OSError as exc:
        raise BridgeError("workspace_config_read_failed", str(exc), {"path": str(path)}) from exc

    try:
        data = json.loads(body)
    except json.JSONDecodeError as exc:
        raise BridgeError("workspace_config_invalid_json", str(exc), {"path": str(path)}) from exc
    if not isinstance(data, dict):
        raise BridgeError("workspace_config_invalid_json", "clock-p.workspace.json must be a JSON object", {"path": str(path)})

    place_id = str(data.get("place_id") or "").strip()
    if not place_id:
        raise BridgeError("workspace_config_missing_place_id", "clock-p.workspace.json is missing place_id", {"path": str(path)})
    if not place_id.isdigit():
        raise BridgeError("workspace_config_invalid_place_id", "clock-p.workspace.json place_id must contain digits only", {"path": str(path), "place_id": place_id})

    code_sync_config = str(data.get("code_sync_config") or "code-sync.roots.json").strip() or "code-sync.roots.json"
    return WorkspaceConfig(place_id=place_id, code_sync_config=code_sync_config)
